sqlmesh/core/analytics/collector.py

```diff
@@ -0,0 +1,351 @@
+from __future__ import annotations
+
+import itertools
+import json
+import typing as t
+from functools import cached_property
+from pathlib import Path
+
+from sqlmesh.core import constants as c
+from sqlmesh.core.analytics.dispatcher import AsyncEventDispatcher, EventDispatcher
+from sqlmesh.utils import random_id
+from sqlmesh.utils.date import now_timestamp
+from sqlmesh.utils.hashing import md5
+from sqlmesh.utils.pydantic import PydanticModel
+from sqlmesh.utils.yaml import dump as yaml_dump
+from sqlmesh.utils.yaml import load as yaml_load
+
+if t.TYPE_CHECKING:
+    from sqlmesh.cicd.config import CICDBotConfig
+    from sqlmesh.core.plan import Plan
+    from sqlmesh.core.snapshot import Snapshot
+
+
+class AnalyticsCollector:
+    def __init__(
+        self,
+        dispatcher: t.Optional[EventDispatcher] = None,
+        sqlmesh_path: t.Optional[Path] = None,
+    ):
+        self._dispatcher = dispatcher or AsyncEventDispatcher()
+        self._sqlmesh_path = sqlmesh_path or c.SQLMESH_PATH
+        self._process_id = random_id()
+        self._seq_num = itertools.count()
+
+    def on_cicd_command(
+        self,
+        *,
+        command_name: str,
+        command_args: t.Collection[str],
+        parent_command_names: t.Collection[str],
+        cicd_bot_config: t.Optional[CICDBotConfig],
+    ) -> None:
+        """Called when a CICD command is executed.
+
+        Args:
+            command_name: The name of the command.
+            command_args: The arguments of the command.
+            parent_command_names: The names of the parent commands.
+            cicd_config: The CICD bot configuration.
+        """
+        additional_args = {}
+        if cicd_bot_config is not None and getattr(cicd_bot_config, "FIELDS_FOR_ANALYTICS", None):
+            additional_args["cicd_bot_config"] = cicd_bot_config.dict(
+                include=cicd_bot_config.FIELDS_FOR_ANALYTICS, mode="json"
+            )
+        self._on_command(
+            "CICD_COMMAND",
+            command_name,
+            command_args,
+            parent_command_names=parent_command_names,
+            **additional_args,
+        )
+
+    def on_cli_command(
+        self,
+        *,
+        command_name: str,
+        command_args: t.Collection[str],
+        parent_command_names: t.Collection[str],
+    ) -> None:
+        """Called when a CLI command is executed.
+
+        Args:
+            command_name: The name of the command.
+            command_args: The arguments of the command.
+            parent_command_names: The names of the parent commands.
```

Comment:
what is a parent command, why does cli have it and not magic?

Reply:
In `click` there's a hierarchy of commands. Eg:
```
sqlmesh plan
```
There are 2 commands here: `sqlmesh` and `plan`. In this case the command name is `plan` and its parent is `sqlmesh.`

With the CICD cli there are 2 levels of commands. Eg:
```
sqlmesh_cicd github run-all
```
`run-all` is the command name, while `sqlmesh_cicd` and `github` are the parents. 

There's no such hierarchy when using magics.